core/src/main/scala/org/apache/spark/deploy/worker/DriverRunner.scala

```diff
@@ -68,56 +70,72 @@ private[deploy] class DriverRunner(
+            kill()
           }
 
-          // TODO: If we add ability to submit multiple jars they should also be added here
-          val builder = CommandUtils.buildProcessBuilder(driverDesc.command, securityManager,
-            driverDesc.mem, sparkHome.getAbsolutePath, substituteVariables)
-          launchDriver(builder, driverDir, driverDesc.supervise)
+          // prepare driver jars, launch driver and set final state from process exit code
+          val exitCode = prepareAndLaunchDriver()
+          finalState = if (exitCode == 0) Some(DriverState.FINISHED) else Some(DriverState.FAILED)
         }
         catch {
-          case e: Exception => finalException = Some(e)
+          case interrupted: InterruptedException =>
+            logInfo("Runner thread for driver " + driverId + " interrupted")
+            killProcessAndFinalize(DriverState.KILLED, interrupted)
+          case e: Exception =>
+            killProcessAndFinalize(DriverState.ERROR, e)
+        }
+        finally {
+          if (shutdownHook != null) ShutdownHookManager.removeShutdownHook(shutdownHook)
         }
 
-        val state =
-          if (killed) {
-            DriverState.KILLED
-          } else if (finalException.isDefined) {
-            DriverState.ERROR
-          } else {
-            finalExitCode match {
-              case Some(0) => DriverState.FINISHED
-              case _ => DriverState.FAILED
-            }
-          }
+        // notify worker of final driver state, possible exception
+        worker.send(DriverStateChanged(driverId, finalState.get, finalException))
+      }
 
+      // kill the process if started, set shared finalizing variables
+      def killProcessAndFinalize(state: DriverState.DriverState, e: Exception): Unit = {
+        killProcess()
         finalState = Some(state)
+        finalException = Some(e)
+      }
+    }
+
+    workerThread.start()
+  }
 
-        worker.send(DriverStateChanged(driverId, state, finalException))
+  /** Kill driver process and wait for it to exit. */
+  private def killProcess(): Unit = {
+    if (process != null) {
+      logInfo("Killing driver process!")
+      val exitCode = Utils.terminateProcess(process, DRIVER_TERMINATE_TIMEOUT_MS)
+      if (exitCode.isEmpty) {
+        logWarning("Failed to terminate driver process: " + process +
+            ". This process will likely be orphaned.")
       }
-    }.start()
+    }
   }
 
-  /** Terminate this driver (or prevent it from ever starting if not yet started) */
-  private[worker] def kill() {
-    synchronized {
-      process.foreach(p => p.destroy())
-      killed = true
+  /** Stop this driver, including the process it launched */
+  private[worker] def kill(): Unit = {
+    if (workerThread != null) {
+      // make sure process does not start if being interrupted
+      this.synchronized {
+        // the workerThread will kill the child process when interrupted
+        workerThread.interrupt()
+        workerThread.join()
```

Comment:
Yeah, your are right, this can cause deadlock.  I'm also having doubts that relying on interrupting the thread to kill the process was a good idea and maybe the current implementation is better.  I'll work on it some more and get these issues ironed out.  Thanks for taking a look!